### PR TITLE
feat: (code actions): add/from destructure

### LINKED ISSUE #167: to destruct, from destruct
- check correct what names would be technically correct
- kind: `refactor.rewrite.add-destruct` and `refactor.rewrite.from-destruct`

Code:

Edge case with to destruct:
```ts
const a = {foo: 5, bar: 8}

const b = [a.foo] // [...] - activation pos range

const c = a
```

```ts
const a = {foo: 5, bar: 8}
const { foo } = a

const b = foo

const c = a
```

With first implementation we can skip hadling this advanced functionality.

### PR DESCRIPTION
**Pull request by Ilanaya**
fixes https://github.com/zardoy/typescript-vscode-plugins/issues/167

[Some tests to play with](https://gist.github.com/maIIady/601de5d400e81231fb1b992694c444b4)

### CODE REVIEW

File: typescript/src/codeActions/custom/addDestructure.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    return true
+}
+
+const isPositionMatchesInitializer = (pos: number, start: number, end: number) => {
+    return pos >= start && pos <= end
+}
+
+export default {
+    id: 'addDestructure',
+    name: 'Add Destructure',
```
Comment (zardoy) [inline typescript/src/codeActions/custom/addDestructure.ts]: wouldnt be better to name to destructure like in linked issue?

**Review by zardoy — CHANGES_REQUESTED** [review]
it does seem to be working only with variable declarations. example to consider:

```ts
const a = ([[{ b, c }]]) => { } // from destruct
```
`[[`, `]]` - selection range. can quickly select with expand selection

**Comment by zardoy** [comment]
from destruct:
the idea is to create a new variable from destruct eg:

```ts
const { a, b } = c
a, b
// after transformation
const newVariable = c
newVariable.a, newVariable.b
```

you can play with actual code actions with the following extension https://drive.google.com/file/d/1ww-EqXaeypy-OhOE5DPUZQ7KjXEvhHJy/view

**Review by Ilanaya — COMMENTED** [review]
It Looks like it is working!

**Review by zardoy — CHANGES_REQUESTED** [review]
I think its good for its purpose.
didn't really look at the code, I think its fine. However, to destructure should also avoid name collisions eg:

```ts
const a = []
const b = {}
b.a
```
from destruct will result in broken code, a should be named as a2 e.g. `a: a2`

**Comment by Ilanaya** [comment]
> I think its good for its purpose. didn't really look at the code, I think its fine. However, to destructure should also avoid name collisions eg:
> 
> ```ts
> const a = []
> const b = {}
> b.a
> ```
> 
> from destruct will result in broken code, a should be named as a2 e.g. `a: a2`

Given this code:
```ts
function fn(newVariable) {
  const something = newVariable.bar
}
```
Source file gonna contain this list of identifiers:
image

So if try getting unique name
```ts
 const uniqueAccessorName = tsFull.getUniqueName('bar', sourceFile as unknown as FullSourceFile)
```
It will *always* generate `bar_1` because it treats `.bar` in access expression `newVariable.bar` as unique identifier.

Hope, this was a clear explanation :) For now I will drop support unique names in this case, or maybe there is some way to handle it?

**Comment by zardoy** [comment]
Thanks for clarifying, we need to remove any `getUniqueName` usage in favor of `typeChecker.resolveName(name, <block body node>, SymbolFlags.Value, /*excludeGlobals*/ true)` (needs full type checker) for checking for name collisions instead (I remember it was also used in extract type, thus need to use type SymbolFlag). Would be possible to make this?

File: typescript/src/utils.ts [hunk 2]
```diff
@@ -312,3 +312,55 @@ export const matchParents: MatchParentsType = (node, treeToCompare) => {
...
+    return true
+}
+export const isNameUniqueAtLocation = (name: string, location: ts.Node | undefined, typeChecker: ts.TypeChecker) => {
+    const checker = getFullTypeChecker(typeChecker)
+
+    const newVariable = checker.resolveName(name, location as unknown as import('typescript-full').Node, ts.SymbolFlags.Value, true)
+    return !newVariable
+}
+export const isNameUniqueAtNodeClosestScope = (name: string, node: ts.Node, typeChecker: ts.TypeChecker) => {
+    const closestScope = findClosestParent(
```
Comment (Ilanaya) [inline typescript/src/utils.ts:369]: Not sure if this is the safest way to get the closest scope where variables should be unique but it works in my cases.

**Comment by Ilanaya** [comment]
> Thanks for clarifying, we need to remove any `getUniqueName` usage in favor of `typeChecker.resolveName(name, <block body node>, SymbolFlags.Value, /*excludeGlobals*/ true)` (needs full type checker) for checking for name collisions instead (I remember it was also used in extract type, thus need to use type SymbolFlag). Would be possible to make this?

Yeah, I'll replace it everywhere in the codebase in a separate PR if you approve the way I implemented this logic here.

File: typescript/src/utils.ts (hunk 2 above)
Comment (zardoy) [reply typescript/src/utils.ts:369]: This is fine I guess, but remember that in es6+ blocks (`{}`) create new scope

Comment (Ilanaya) [reply typescript/src/utils.ts:369]: fixed + added test

**Comment by zardoy** [comment]
This is getting great! So excited to see it in repo, @Ilanaya is it ready to merge now?

**Comment by Ilanaya** [comment]
Yep, I think it is. I'll recheck it again soon and merge if u don't mind

**Comment by zardoy** [comment]
@Ilanaya  ok, I've got time to check how it works and I must admit there are some improvements needed:

1. need to handle case if property is reserved (e.g. `class` or `default` might be pretty common):

```ts
const a = {
    default: 1,
}

a.default
// results in broken code:
const { class } = { // class is resorved token!
    class: 1,
}

class
```
If you use require in js file you will have a quickfix to convert the file to es modules eg:

```js
module.exports.default = 5
```
will be converted to
```js
const _default = 5
export { _default as default }
```
it just uses isIdentifierANonContextualKeyword: https://github.com/a-tarasyuk/typescript/blob/8dc7ee8c2b3eadcd4d1130afb9c3a9df927ef6a7/src/services/codefixes/convertToEsModule.ts#L165

2. And also, have we ever discussed how we should handle index access? e.g.

```ts
const a = {
    a: 1,
}

a.a
a['b']
a['c-d']
```

**Pull request merged**